File: python/src/nitroping/_notifications.py
"""``notifications`` resource client.

Mounted on :class:`nitroping.Nitroping` as ``np.notifications``. Wraps
``POST /api/v1/notifications`` and ``GET /api/v1/notifications/:id``.
"""

from __future__ import annotations

from typing import Any, cast
from urllib.parse import quote

from ._http import HttpClient
from .types import (
    CancelNotificationResult,
    NotificationAction,
    NotificationResult,
    NotificationTarget,
)


def _target_to_wire(target: NotificationTarget) -> dict[str, Any]:
    """Normalize a target selector for the wire.

    The wire shape is the same snake_case dict the caller passes; the
    only transform is defaulting a segment target's ``match`` to
    ``"all"`` when the caller omits it.
    """
    wire = dict(target)
    segment = wire.get("segment")
    if isinstance(segment, dict):
        normalized = dict(segment)
        normalized.setdefault("match", "all")
        wire["segment"] = normalized
    return wire


class NotificationsClient:
    """Send and inspect notifications."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def send(
        self,
        *,
        target: NotificationTarget,
        title: str | None = None,
        body: str | None = None,
        template: str | None = None,
        vars: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
        icon: str | None = None,
        image: str | None = None,
        click_action: str | None = None,
        deep_link: str | None = None,
        actions: list[NotificationAction] | None = None,
        apns_category: str | None = None,
        scheduled_at: str | None = None,
        expires_at: str | None = None,
        recurrence: str | None = None,
        recurrence_tz: str | None = None,
        recurrence_until: str | None = None,
        email_to: list[str] | None = None,
        idempotency_key: str | None = None,
    ) -> NotificationResult:
        """Enqueue a new notification.

        Either ``title + body`` (raw payload) or ``template + vars``
        (Pro plan). Mixing the two is a 422.

        ``apns_category`` is iOS only. It sets ``aps.category`` verbatim
        so an app that registered a matching ``UNNotificationCategory``
        renders the action buttons. Overrides the server-minted category
        for this message.

        ``recurrence`` (a 5-field cron string) with optional
        ``recurrence_tz`` and ``recurrence_until`` schedules a repeating
        send. ``email_to`` adds email recipients for the notification.

        Returns ``{"id": ..., "status": ...}`` on ``201 Created``. On
        non-2xx the SDK raises :class:`~nitroping.errors.ApiError`
        carrying the server's ``code``, ``message``, and (for validation
        failures) the per-field ``details`` map.
        """
        wire: dict[str, Any] = {"target": _target_to_wire(target)}
        if title is not None:
            wire["title"] = title
        if body is not None:
            wire["body"] = body
        if template is not None:
            wire["template"] = template
        if vars is not None:
            wire["vars"] = vars
        if data is not None:
            wire["data"] = data
        if icon is not None:
            wire["icon"] = icon
        if image is not None:
            wire["image"] = image
        if click_action is not None:
            wire["click_action"] = click_action
        if deep_link is not None:
            wire["deep_link"] = deep_link
        if actions is not None:
            wire["actions"] = actions
        if apns_category is not None:
            wire["apns_category"] = apns_category
        if scheduled_at is not None:
            wire["scheduled_at"] = scheduled_at
        if expires_at is not None:
            wire["expires_at"] = expires_at
        if recurrence is not None:
            wire["recurrence"] = recurrence
        if recurrence_tz is not None:
            wire["recurrence_tz"] = recurrence_tz
        if recurrence_until is not None:
            wire["recurrence_until"] = recurrence_until
        if email_to is not None:
            wire["email_to"] = email_to

        headers: dict[str, str] = {}
        if idempotency_key is not None:
            headers["Idempotency-Key"] = idempotency_key

        response = self._http.request(
            "POST", "/api/v1/notifications", body=wire, headers=headers
        )
        return cast(NotificationResult, response)

    def get(self, notification_id: str) -> dict[str, Any]:
        """Fetch a previously-enqueued notification by id.

        Returns the full row (including counters: ``total_sent``,
        ``total_delivered``, ``total_failed``, etc.).
        """
        response = self._http.request(
            "GET", f"/api/v1/notifications/{quote(notification_id, safe='')}"
        )
        return cast(dict[str, Any], response)

    def cancel(self, notification_id: str) -> CancelNotificationResult:
        """Cancel a scheduled or in-flight notification.

        Wraps ``DELETE /api/v1/notifications/:id``. Returns
        ``{"id": ..., "status": "canceled"}``. Raises
        :class:`~nitroping.errors.ApiError` with ``code = "cannot_cancel"``
        (409) if the notification already reached a terminal state, or
        ``code = "not_found"`` (404).
        """
        response = self._http.request(
            "DELETE", f"/api/v1/notifications/{quote(notification_id, safe='')}"
        )
        return cast(CancelNotificationResult, response)

File: kotlin/nitroping/src/main/kotlin/dev/nitroping/Notifications.kt
/*
 * Notifications.kt
 *
 * `np.notifications` — wraps `POST /api/v1/notifications` and
 * `GET /api/v1/notifications/:id`.
 *
 * Mostly used from Kotlin-on-the-server (Ktor, Spring); Android apps
 * receive notifications, they don't typically send them.
 */

package dev.nitroping

import dev.nitroping.internal.HttpTransport

public class Notifications internal constructor(private val transport: HttpTransport) {
    /**
     * Enqueue a notification.
     *
     * Returns [NotificationResult] on `201 Created`. On non-2xx the SDK
     * throws [ApiException] carrying the server's `code`, `message`, and
     * (for validation failures) the per-field `details` map.
     */
    public suspend fun send(input: SendRequest): NotificationResult {
        val headers: Map<String, String> = input.idempotencyKey
            ?.let { mapOf("Idempotency-Key" to it) }
            ?: emptyMap()
        val raw = transport.request(
            method = "POST",
            path = "/api/v1/notifications",
            body = toWire(input),
            headers = headers,
        )
        return decodeResult(raw)
    }

    /**
     * Fetch a previously enqueued notification by id.
     *
     * Returns the raw JSON tree (`Map<String, Any?>`) so callers can read
     * the full row including counters (`total_sent`, `total_delivered`,
     * etc.) without forcing the SDK to model every future field.
     */
    @Suppress("UNCHECKED_CAST")
    public suspend fun get(id: String): Map<String, Any?> {
        require(id.isNotEmpty()) { "Notification id must not be empty" }
        val raw = transport.request("GET", "/api/v1/notifications/$id")
        return (raw as? Map<String, Any?>)
            ?: throw NitropingException("Unexpected response shape for GET notification", code = "decode_error")
    }

    /**
     * Cancel a queued or scheduled notification.
     *
     * Returns `{ id, status = "canceled" }`. Throws [ApiException] with
     * `code = "not_found"` if the id doesn't belong to your app, or
     * `code = "cannot_cancel"` (409) if the notification already reached a
     * terminal state (sent / failed / partial / canceled).
     */
    public suspend fun cancel(id: String): NotificationCancelResult {
        require(id.isNotEmpty()) { "Notification id must not be empty" }
        val raw = transport.request("DELETE", "/api/v1/notifications/$id")
        val map = raw as? Map<*, *>
            ?: throw NitropingException("Unexpected response shape", code = "decode_error")
        return NotificationCancelResult(
            id = map["id"] as? String
                ?: throw NitropingException("Missing `id` in response", code = "decode_error"),
            status = map["status"] as? String
                ?: throw NitropingException("Missing `status` in response", code = "decode_error"),
        )
    }

    /** camelCase → snake_case + drop `null`s, matching the JS SDK's `toWire`. */
    private fun toWire(input: SendRequest): Map<String, Any?> {
        val wire = LinkedHashMap<String, Any?>()
        input.title?.let { wire["title"] = it }
        input.body?.let { wire["body"] = it }
        input.template?.let { wire["template"] = it }
        input.vars?.let { wire["vars"] = it }
        input.data?.let { wire["data"] = it }
        input.icon?.let { wire["icon"] = it }
        input.image?.let { wire["image"] = it }
        input.clickAction?.let { wire["click_action"] = it }
        input.deepLink?.let { wire["deep_link"] = it }
        input.actions?.let { list ->
            wire["actions"] = list.map { a ->
                val m = LinkedHashMap<String, Any?>()
                m["id"] = a.id
                m["title"] = a.title
                if (a.icon != null) m["icon"] = a.icon
                m
            }
        }
        input.apnsCategory?.let { wire["apns_category"] = it }
        input.scheduledAt?.let { wire["scheduled_at"] = it }
        input.expiresAt?.let { wire["expires_at"] = it }
        input.recurrence?.let { wire["recurrence"] = it }
        input.recurrenceTz?.let { wire["recurrence_tz"] = it }
        input.recurrenceUntil?.let { wire["recurrence_until"] = it }
        input.emailTo?.let { wire["email_to"] = it }
        wire["target"] = targetToWire(input.target)
        return wire
    }

    private fun targetToWire(target: Target): Map<String, Any?> = when (target) {
        Target.All -> mapOf("all" to true)
        is Target.DeviceIds -> mapOf("device_ids" to target.ids)
        is Target.UserIds -> mapOf("user_ids" to target.ids)
        is Target.Tags -> mapOf("tags" to target.tags)
        is Target.Segment -> mapOf(
            "segment" to mapOf(
                "match" to target.match,
                "conditions" to target.conditions.map { c ->
                    val m = LinkedHashMap<String, Any?>()
                    m["field"] = c.field
                    m["op"] = c.op
                    if (c.value != null) m["value"] = c.value
                    m
                },
            ),
        )
    }

    private fun decodeResult(raw: Any?): NotificationResult {
        val map = raw as? Map<*, *>
            ?: throw NitropingException("Unexpected response shape", code = "decode_error")
        val id = map["id"] as? String
            ?: throw NitropingException("Missing `id` in response", code = "decode_error")
        val status = map["status"] as? String
            ?: throw NitropingException("Missing `status` in response", code = "decode_error")
        return NotificationResult(id = id, status = status)
    }
}
